torchtext/csrc/sentencepiece.cpp

```diff
@@ -0,0 +1,89 @@
+#include <sentencepiece_trainer.h>
+#include <torch/script.h>
+
+namespace torchtext {
+namespace {
+
+struct SentencePiece : torch::CustomClassHolder {
+private:
+  sentencepiece::SentencePieceProcessor processor_;
+
+public:
+  std::string content_;
+  explicit SentencePiece(const std::string &content) : content_(content) {
```

Comment:
This could be changed to take a ::sentencepiece::<Sentence piece class> via move constructor and then effectively represent a shell that forwards methods to the underlying object.

Do you need this string constructor? Does this match previous behavior? Usually I'd make this a separate factory function and then construct a SentencePiece object that is initialized with whatever ```processor_.LoadFromSerializedProto(content_);``` returns.

Reply:
This class has to provide serialization mechanism so that scripted module can be saved.
Ideally, the underlying `SentencePieceProcessor` class should provide such mechanism, but it does not provide interface to serialize the model as protobuf.
Note 1. The lifecycle of the official `SentencePieceProcessor` is 1. train and save to a file then 2. load from the saved file. It does not have a functionality to save the model.
Note 2. If you look at [the source code of `SentencePieceProcessor`](https://github.com/google/sentencepiece/blob/master/src/sentencepiece_processor.h#L462), it almost look like provide serialization interface, but `ModelProto` and all the functionality related to `protobuf` does not exist in the library file.

To workaround this, we need to keep the content of the saved model file (model data serialized protobuf) when loading a model from a file.
So, it is simpler to pass the model content as string to constructor and to instantiate the underlying `SentencePieceProcessor` there. Because, you still need to pass the model content separately for the later serialization.

Reply:
The serialization here is necessary to support pickle? As far as I know that's what we use to save models. We could look into adding custom pickle support [as does pybind11](https://pybind11.readthedocs.io/en/stable/advanced/classes.html#pickling-support) for this. This might also enable us, if they don't already or if there's a need, to apply compression algorithms on the fly and save a smaller model.